processor/src/main/java/org/mapstruct/ap/model/BeanMappingMethod.java

```diff
@@ -121,17 +126,66 @@ public BeanMappingMethod build() {
                 if ( beanMapping != null && beanMapping.getResultType() != null ) {
                     resultType = ctx.getTypeFactory().getType( beanMapping.getResultType() );
                     if ( !resultType.isAssignableTo( method.getResultType() ) ) {
-                        ctx.getMessager().printMessage( method.getExecutable(),
+                        ctx.getMessager().printMessage(
+                            method.getExecutable(),
                             beanMappingPrism.mirror,
-                            Message.BEANMAPPING_NOT_ASSIGNABLE, resultType, method.getResultType());
+                            Message.BEANMAPPING_NOT_ASSIGNABLE, resultType, method.getResultType()
+                        );
                     }
                 }
             }
 
+            sortPropertyMappingsByDependencies();
+
             return new BeanMappingMethod( method, propertyMappings, factoryMethod, mapNullToDefault, resultType );
         }
 
         /**
+         * Sources the given mappings as per the dependency relationships given via {@code dependsOn()}. If a cycle is
+         * detected, an error is reported.
+         */
+        private void sortPropertyMappingsByDependencies() {
+            final GraphAnalyzer graphAnalyzer = new GraphAnalyzer();
+
+            for ( PropertyMapping propertyMapping : propertyMappings ) {
+                graphAnalyzer.addNode( propertyMapping.getName(), propertyMapping.getDependsOn() );
+            }
+
+            graphAnalyzer.analyze();
+
+            if ( !graphAnalyzer.getCycles().isEmpty() ) {
+                Set<String> cycles = new HashSet<String>( graphAnalyzer.getCycles().size() );
```

Comment:
Setting this initial capacity will result in a resizing, as the default load factor is 0.75. In order to avoid the resizing, you could use something like ```Math.ceil( (float) expectedSize * 4.f / 3.f )``` - or take a look at Maps.newHashMapWithExpectedSize in guava... :wink: